scio-test/src/it/scala/com/spotify/ScioBenchmark.scala

```diff
@@ -377,8 +417,136 @@ private[this] object PrettyPrint {
 // sbt scio-test/it:runMain com.spotify.ScioBatchBenchmarkResult $buildNum1 $buildNum2
 // where $buildNum1 and $buildNum2 are build number of "bench" jobs in CircleCI
 object ScioBatchBenchmarkResult {
+
   import ScioBenchmarkSettings._
+
   def main(args: Array[String]): Unit =
     new DatastoreLogger(BatchMetrics)
       .printMetricsComparison(ScioBatchBenchmark.benchmarkNames, Some((args(0), args(1))))
 }
+
+abstract class Benchmark(val extraConfs: Map[String, Array[String]] = Map.empty)
+    extends Serializable {
+  val name: String = this.getClass.getSimpleName.replaceAll("\\$$", "")
+
+  private val configurations: Map[String, Array[String]] = {
+    val base = Map(name -> Array.empty[String])
+    val extra = if (extraConfs == null) {
+      Map.empty
+    } else {
+      extraConfs.map(kv => (s"$name${kv._1}", kv._2))
+    }
+    base ++ extra
+  }
+
+  def run(projectId: String,
+          prefix: String,
+          args: Array[String]): Iterable[Future[BenchmarkResult]] = {
+    val username = CoreSysProps.User.value
+    configurations
+      .map {
+        case (confName, extraArgs) =>
+          val (sc, _) =
+            ContextAndArgs(Array(s"--project=$projectId") ++ args ++ extraArgs)
+          sc.setAppName(confName)
+          sc.setJobName(s"$prefix-$confName-$username".toLowerCase())
+          run(sc)
+          val result = sc.close()
+          result.finalState.map(_ => BenchmarkResult.batch(confName, extraArgs, result))
+      }
+  }
+
+  def run(sc: ScioContext): Unit
+
+  def randomUUIDs(sc: ScioContext, n: Long): SCollection[Elem[String]] =
+    sc.parallelize(partitions(n)).transform("UUID-generator") {
+      _.flatten[Long]
+        .applyTransform(ParDo.of(new FillDoFn(() => UUID.randomUUID().toString)))
+        .map(Elem(_))
+    }
+
+  def randomKVs(sc: ScioContext, n: Long, numUniqueKeys: Int): SCollection[(String, Elem[String])] =
+    sc.parallelize(partitions(n))
+      .flatten[Long]
+      .applyTransform(ParDo.of(new FillDoFn(() =>
+        ("key" + Random.nextInt(numUniqueKeys), UUID.randomUUID().toString))))
+      .mapValues(Elem(_))
+
+  def withRandomKey[T: Coder](n: Int): SCollection[T] => SCollection[(Int, T)] =
+    _.keyBy(_ => Random.nextInt(n))
+
+  private def partitions(n: Long,
+                         numPartitions: Int = 100,
+                         numOfWorkers: Int = numOfWorkers): Iterable[Iterable[Long]] = {
+    val chunks = numPartitions * numOfWorkers
+
+    def loop(n: Long): Seq[Long] = {
+      n match {
+        case 0                    => Nil
+        case x if x < chunks      => Seq(x)
+        case x if x % chunks == 0 => Seq.fill(chunks)(x / chunks)
+        case x =>
+          val r = x % chunks
+          loop(r) ++ loop(x - r)
+      }
+    }
+
+    loop(n).grouped(numOfWorkers).toIterable
+  }
+
+  private class FillDoFn[T](val f: () => T) extends DoFn[Long, T] {
```

Comment:
@andrisnoko all the above methods should including this class should be moved to either the companion object `Benchmark`